lib: Add methods to clean up refs in local repository

smspillaz opened the pull request:
Add flatpak_installation_remove_local_ref_sync to remove a given ref from the local repository if the ref is known and flatpak_installation_cleanup_local_refs_sync to remove all undeployed refs.

This changes also adds a RemoveLocalRef helper method to the system helper, since modifying the system OSTree repo is a privileged operation. Removing a local ref probably does not impose any notable security risks, since the worst that could happen is a DoS where the user would have to download it again. Theoretically, it could be used in an attack where the local refs are removed and refs are re-downloaded from a compromised OSTree remote, but if that occurs there are much bigger problems (the remote owner would have lost control of their keys).

Fixes #1031

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7012,6 +7012,58 @@ flatpak_dir_cleanup_removed (FlatpakDir   *self,
+  g_hash_table_iter_init (&hash_iter, local_refs);
+  while (g_hash_table_iter_next (&hash_iter, &key, NULL))
+    {
+      g_autoptr(GVariant) deploy_data = flatpak_dir_get_deploy_data (self, key, NULL, NULL);
```
This will filter out all appstream refs, which is not great.

alexlarsson commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,59 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * pulled a flatpak ref using flatpak_installation_install_full and
+ * specified FLATPAK_INSTALL_FLAGS_NO_DEPLOY but then decided not to
+ * deploy the ref later on and want to remove the local ref to prevent it
+ * from taking up disk space.
```
Maybe we also need a prune operation?

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -6953,10 +6953,36 @@ flatpak_dir_remove_ref (FlatpakDir   *self,
     return FALSE;
 
-  return TRUE;
+  return flatpak_dir_prune (self, cancellable, error);
```
This will always prune, which is already done in e.g. flatpak_dir_uninstall() which calls remove_ref(). This means it will be done twice, and additionally, with the new prune locks, i think this will deadlock.

Also, i think it will not be entirely uncommon to remove several refs and in that case it will be more useful to prune only at the end.

smspillaz replied:
Yeah. I think what I will do here is pull the `flatpak_dir_prune` to the library boundary (so if you remove a single local ref, the prune happens immediately afterwards, otherwise the prune happens once all refs are removed).

Or should we make it the caller's responsibility and expose it in the library?

alexlarsson commented on system-helper/flatpak-system-helper.c:
```diff
@@ -791,6 +791,49 @@ handle_update_remote (FlatpakSystemHelper *object,
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
```
Maybe this should fail if the specified ref is already deployed at the same version? It seems weird to allow the system repo to end up in a somewhat inconsistent state.

smspillaz replied:
I'm not sure I understand what is meant here. If you are saying that we should not be permitted to remove refs from the local repo if they are already deployed, I'm not sure what impact that would have - I suppose the user would to re-pull the ref. I suppose what I could do is add a check to `flatpak_dir_remove_ref` that checks if the ref is deployed for now.

smspillaz replied:
Hmm, actually, I don't think that will work. It is possible to remove a remote and at that point all refs are removed from the local repo. This would put the repo in the same "inconsistent state".

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7550,13 +7576,19 @@ flatpak_dir_remote_list_refs (FlatpakDir       *self,
 
+typedef enum {
+  FIND_MATCHING_REFS_FLAGS_NONE = 0,
+  FIND_MATCHING_REFS_FLAGS_KEEP_REMOTE = (1 << 0)
```
Nitpick: Add a trailing comma on this one to avoid diff noise in future if adding more enum members.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8032,80 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
 }
 
+static GPtrArray *
+filter_out_deployed_refs (FlatpakDir *self,
```
Nitpick: It would be good to have a comment above this function summarising what it does; makes the code a bit easier to understand at a glance.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8032,80 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+}
+
+gboolean
+flatpak_dir_cleanup_undeployed_refs (FlatpakDir   *self,
```
Nitpick: It would be good to have a comment above this function summarising what it does; makes the code a bit easier to understand at a glance.

cosimoc commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8032,80 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+
+      if (!ostree_parse_refspec (refspec, &remote, &ref, error))
```
I think you can pass `NULL` as the remote, since it's not used.

cosimoc commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8032,80 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+
+  for (; i < undeployed_refs->len; ++i)
+    {
+      const char *key = g_ptr_array_index (undeployed_refs, i);
```
Nitpick: use `refspec` for this variable name

cosimoc commented on system-helper/flatpak-system-helper.c:
```diff
@@ -791,6 +791,49 @@ handle_update_remote (FlatpakSystemHelper *object,
+                         GDBusMethodInvocation *invocation,
+                         const gchar *arg_remote,
+                         const gchar *arg_ref,
+                         const gchar *arg_installation)
```
Alignment

smspillaz commented:
Okay, rebased and updated this. Main changes are:
 - Added system helper for flatpak_dir_prune
 - Added `flatpak_installation_prune_local_repo`
 - Fixed nitpiks
 - Added more gtkdoc

I decided not to include changes made locally to fail removing a ref that is already deployed - this already happens in the case where a repo is removed.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -6946,14 +6946,57 @@ flatpak_dir_undeploy_all (FlatpakDir   *self,
+ * OSTree repo. Attempting to remove a ref that is currently deployed
+ * is an error, you need to uninstall the flatpak first. Note that this does
+ * not remove the objects bound to @ref from the disk, you will need to
+ * call flatpak_dir_prune to do that.
```
Nitpick: add `()` after `flatpak_dir_prune` (no space between them) and gtk-doc will automatically add a cross-reference link to the function.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -6946,14 +6946,57 @@ flatpak_dir_undeploy_all (FlatpakDir   *self,
+ * @remote_name: the name of the remote
+ * @ref: the flatpak ref to remove
+ * @cancellable: (allow-none): a #GCancellable
+ * @error: (allow-none): a #GError
```
Argument documentation goes at the top of gtk-doc comments, immediately after the function name.

pwithnall replied:
Drop the `(allow-none)` annotation from `@error`. It’s not needed (`g-ir-scanner` already knows about the conventions for `GError` arguments) and is deprecated. In this case, in particular, `(allow-none)` would be replaced by `(nullable) (optional)`, thus demonstrating the difference between the two replacements for `(allow-none)`. `(nullable)` on an out argument means `NULL` may be passed out; `(optional)` on an out argument means `NULL` may be passed in.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -6946,14 +6946,57 @@ flatpak_dir_undeploy_all (FlatpakDir   *self,
+ * @self: a #FlatpakDir
+ * @remote_name: the name of the remote
+ * @ref: the flatpak ref to remove
+ * @cancellable: (allow-none): a #GCancellable
```
`(allow-none)` is deprecated and has been replaced by `(nullable)`. See https://wiki.gnome.org/Projects/GObjectIntrospection/Annotations#Nullable_parameters.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+{
+  g_autoptr(GPtrArray) undeployed_refs = g_ptr_array_new_full (local_refspecs->len, g_free);
+  g_autoptr(GVariant) deploy_data = NULL;
+  gsize i = 0;
```
Move the initialisation of `i` into the `for` loop to make the bounds more obvious.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+ *
+ * @self: a #FlatpakDir
+ * @cancellable: (allow-none): a #GCancellable
+ * @error: (allow-none): a #GError
```
See previous comments about `(allow-none)` and about putting argument documentation at the top of the gtk-doc comment.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+ * @error: (allow-none): a #GError
+ *
+ * Since: 0.10.0
+ * Returns: (transfer-full): %TRUE if cleaning up the refs suceeded, %FALSE
```
`(transfer full)` rather than `(transfer-full)` (space not hyphen).

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+/**
+ * flatpak_dir_cleanup_undeployed_refs:
+ *
+ * Find all flatpak refs in the local repository which have not been deloyed
```
s/deloyed/deployed/ (typo)

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+ * function if you pulled refs into the dir but then decided that you did
+ * not want to deploy them for some reason. Note that this does not prune
+ * objects bound to the cleaned up refs from the underlying OSTree repository,
+ * you should consider using flatpak_dir_prune to do that.
```
Nitpick: Add `()` to `flatpak_dir_prune()`.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+  g_autoptr(GHashTable) local_refspecs = NULL;
+  g_autoptr(GPtrArray)  local_flatpak_refspecs = NULL;
+  g_autoptr(GPtrArray) undeployed_refs = NULL;
+  gsize i = 0;
```
Move the initialisation of `i` into the `for` loop.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
+gboolean
+flatpak_dir_cleanup_undeployed_refs (FlatpakDir   *self,
+                                     GCancellable *cancellable,
+                                     GError       **error)
```
Nitpick: argument alignment.

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * Remove the OSTree ref given by @remote_name:@ref from the local flatpak
+ * repository. The next time the underlying OSTree repo is pruned, objects
+ * which were attached to that ref will be removed. This is useful if you
+ * pulled a flatpak ref using flatpak_installation_install_full and
```
Nitpick: add `()` to `flatpak_installation_install_full()`.

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * repository. The next time the underlying OSTree repo is pruned, objects
+ * which were attached to that ref will be removed. This is useful if you
+ * pulled a flatpak ref using flatpak_installation_install_full and
+ * specified FLATPAK_INSTALL_FLAGS_NO_DEPLOY but then decided not to
```
Nitpick: Prefix `FLATPAK_INSTALL_FLAGS_NO_DEPLOY` with `%` to get gtk-doc to format it appropriately.

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * deploy the ref later on and want to remove the local ref to prevent it
+ * from taking up disk space. Note that this will not remove the objects
+ * referred to by @ref from the underlying OSTree repo, you should use
+ * flatpak_installation_prune_local_repo to do that.
```
Nitpick: `()`

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * referred to by @ref from the underlying OSTree repo, you should use
+ * flatpak_installation_prune_local_repo to do that.
+ *
+ * Returns: %TRUE on success
```
Nitpick: Missing a `Since: …` line

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * referred to by @ref from the underlying OSTree repo, you should use
+ * flatpak_installation_prune_local_repo to do that.
+ *
+ * Returns: %TRUE on success
```
Nitpick: Missing a `Since: …` line

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * deploy the refs later on and want to remove the local refs to prevent them
+ * from taking up disk space. Note that this will not remove the objects
+ * referred to by @ref from the underlying OSTree repo, you should use
+ * flatpak_installation_prune_local_repo to do that.
```
See documentation comments/nitpicks above.

pwithnall commented on lib/flatpak-installation.c:
```diff
@@ -2041,3 +2041,84 @@ flatpak_installation_list_installed_related_refs_sync (FlatpakInstallation *self
+ * Remove all orphaned OSTree objects from the underlying OSTree repo in
+ * @installation.
+ *
+ * Returns: %TRUE on success
```
Nitpick: Missing a `Since: …` line

smspillaz commented:
Remaining nitpicks should be addressed

alexlarsson commented:
@rh-atomic-bot r+ c45df7d5643c7212bd22470f36ed8e60bb10f38b

alexlarsson commented:
Failed due to unused variable..

smspillaz commented:
@alexlarsson should be fixed in case you want to take another look

alexlarsson commented:
@rh-atomic-bot r+
